17013 - migrate OptionRow to PressableWithFeedback

Skalakid opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ [17013](https://github.com/Expensify/App/issues/17013)
PROPOSAL: [17013](https://github.com/Expensify/App/issues/17013)


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

This component is used in many places, you can test it by checking the list of users that
-  reacted with specific emoji
- are members of a workspace
- are in the group chat
- in your workspace (list with the invite and remove buttons)

Steps that I took:
1. Log in
2. Enter a group chat
3. Click its name, to see its details
4. Click `Members`
5. Now you should see the list of users that uses the `OptionRow` component. Verify if these buttons work fine

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
6. Upload an image via copy paste
7. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/39538890/25743c34-f9de-4816-89de-07303ca7972a


</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/39538890/0c74c541-571a-4912-83e5-3bd387fad9cd


</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/39538890/d6fa2f36-ba51-4b1f-bb48-aa628b7ceae4


</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/39538890/a88d31b1-8725-408a-99a5-b764f33cfc9f



</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/39538890/e490fb61-f5f6-4be3-ac7f-f57cc6290a52


</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/39538890/87e2d1e2-bf48-4dd8-a4df-fd4513465adf


</details>

Skalakid committed:
migrate OptionRow to PressableWithFeedback

Skalakid committed:
fixed dimming

Skalakid committed:
removed comment

arosiclair commented:
@Skalakid make sure to assign yourself when creating PRs. Some of our automation depends on that. @robertKozik could you do the checklist for this as well?

robertKozik commented:
Sure, I'll do it later today

robertKozik commented on src/components/OptionRow.js:
```diff
@@ -174,6 +174,10 @@ class OptionRow extends Component {
+                            accessibilityLabel={this.props.option.text}
+                            accessibilityRole="button"
+                            hoverDimmingValue={1}
+                            pressDimmingValue={0.8}
```
the current default value for press opacity is the same
https://github.com/Expensify/App/blob/79ddcab5c118b2b5c3769150d3177b60bb58e86c/src/styles/variables.js#L143

robertKozik commented on src/components/OptionRow.js:
```diff
@@ -174,6 +174,10 @@ class OptionRow extends Component {
                                 this.props.isDisabled && styles.cursorDisabled,
```
I know it wasn't changed in your PR, but this style is now redundant as `PressableWithFeedback` is managing it itself. Thus this line can be removed.

robertKozik reviewed (changes requested):
Please change these two thing and It should be GTG

Skalakid committed:
review changes

Skalakid committed:
removed unnecessary changes

robertKozik commented:
@Skalakid  I've found one inconsistency. With the new implementation if the optionRow is in keyboard focus and then hovered the blue border is dismissed. In the old one it stayed
steps to repro:
1. Enter a group chat
2. Click its name, to see its details
3. Click Members
4. With `tab` key focus the optionRow
5. Hover focused row with mouse

CC. @arosiclair
<details>
<summary>OLD</summary>

https://github.com/Expensify/App/assets/61146594/22b3d443-fd73-4f97-9842-642400cc06de


</details>

<details>
<summary>NEW</summary>

https://github.com/Expensify/App/assets/61146594/f789f620-6905-4f5c-affc-f158cc8c0293


</details>

Skalakid committed:
fixed styling on focus

Skalakid commented:
@arosiclair I've added a fix for the on-focus styling problem, that @robertKozik mentioned above. To prevent the blue border from disappearing I changed the focus style to be the same as the hover style. With this focused style looks more standardized, what do you think about it?

arosiclair commented:
Hmm I think preferably we just apply both the focus border and hover highlight. It looks like `PressibleWithFeedback` [_should_ be able to apply both](https://github.com/Expensify/App/blob/8e6c3e20142a797669aabb8ff4b499c29a66063d/src/components/Pressable/PressableWithFeedback.js#L59-L64). Why isn't it working here?

Skalakid commented:
> Hmm I think preferably we just apply both the focus border and hover highlight. It looks like `PressibleWithFeedback` [_should_ be able to apply both](https://github.com/Expensify/App/blob/8e6c3e20142a797669aabb8ff4b499c29a66063d/src/components/Pressable/PressableWithFeedback.js#L59-L64). Why isn't it working here?

The blue border is rendered by a web browser, I believe. We are adding styles to `OpacityView` so they are rendering above the blue border. I think we can apply both blue border and custom styles by passing them to the `GenericPressable` and using `hoverStyle` and `focusStyle`. However, it can limit our styling possibilities because in this situation we would have only one child - `OpacityView`, not the ones that we passed.
But I've replaced the blue border in #19660  and I got design acceptance ([comment](https://github.com/Expensify/App/pull/19660#issuecomment-1572012196)) so I think it can be like that :D

Skalakid committed:
Merge branch 'main' into 17013-migrate-OptionRow

robertKozik commented:
Looks good 🚀  Reviewer checklist is now filled  @arosiclair all yours

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5213535458) to staging by https://github.com/arosiclair in version: 1.3.26-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

isagoico commented:
This PR QA is failing because of this issue https://github.com/Expensify/App/issues/20513

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5249654407) to production by https://github.com/AndrewGable in version: 1.3.26-4 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

Christinadobrzyn commented:
This might be a regression - https://github.com/Expensify/App/issues/20585

@Skalakid @arosiclair could you take a peek?

s77rt commented on src/components/OptionRow.js:
```diff
@@ -176,6 +176,7 @@ class OptionRow extends Component {
                             accessibilityRole="button"
                             hoverDimmingValue={1}
                             hoverStyle={this.props.hoverStyle}
+                            focusStyle={this.props.hoverStyle}
```
This caused a regression https://github.com/Expensify/App/issues/20585
